cmd/localapiclient/localapiclient.go

```diff
@@ -1,87 +1,82 @@
 package localapiclient
 
 import (
-	"bytes"
-	"errors"
+	"context"
 	"fmt"
-	"log"
+	"io"
 	"net/http"
 	"time"
 
 	"tailscale.com/ipn/localapi"
 )
 
-// LocalAPIResponseWriter substitutes for http.ResponseWriter in order to write byte streams directly
-// to a receiver function in the application.
-type LocalApiResponseWriter struct {
-	headers http.Header
-	body    bytes.Buffer
-	status  int
+// Response represents the result of processing an http.Request.
+type Response struct {
+	headers    http.Header
+	status     int
+	bodyWriter io.WriteCloser
+	bodyReader io.ReadCloser
 }
 
-func newLocalApiResponseWriter() *LocalApiResponseWriter {
-	return &LocalApiResponseWriter{headers: http.Header{}, status: http.StatusOK}
-}
-
-func (w *LocalApiResponseWriter) Header() http.Header {
-	return w.headers
+func (r *Response) Header() http.Header {
+	return r.headers
 }
 
 // Write writes the data to the response body, which will be sent to Java. If WriteHeader is not called
 // explicitly, the first call to Write will trigger an implicit WriteHeader(http.StatusOK).
-func (w *LocalApiResponseWriter) Write(data []byte) (int, error) {
-	if w.status == 0 {
-		w.WriteHeader(http.StatusOK)
+func (r *Response) Write(data []byte) (int, error) {
+	if r.status == 0 {
+		r.WriteHeader(http.StatusOK)
 	}
-	return w.body.Write(data)
+	return r.bodyWriter.Write(data)
 }
 
-func (w *LocalApiResponseWriter) WriteHeader(statusCode int) {
-	w.status = statusCode
+func (r *Response) WriteHeader(statusCode int) {
+	r.status = statusCode
 }
 
-func (w *LocalApiResponseWriter) Body() []byte {
-	return w.body.Bytes()
+func (r *Response) Body() io.ReadCloser {
```

Comment:
Note that this changes the return type from `[]byte` to `io.ReadCloser`. This is less convenient for the caller because it requires them to read from the reader, and it requires them to close the reader when finished, but it has the advantage of not having to read the whole response body into memory if we don't want to or have to.

Reply:
Note also that io.ReadCloser won't be able to be bound to Swift, so to maintain similar streaming semantics we'll probably need to get creative with our own interface using callbacks or something like that.

Reply:
This is definitely an improvement, thanks!